MonoGame.Framework/Utilities/Png/PngReader.cs

```diff
@@ -0,0 +1,331 @@
+﻿// MonoGame - Copyright (C) The MonoGame Team
+// This file is subject to the terms and conditions defined in
+// file 'LICENSE.txt', which is part of this source code package.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.Specialized;
+using System.IO;
+using System.Linq;
+using System.Text;
+using MonoGame.Utilities.ZLib;
+using Microsoft.Xna.Framework.Graphics;
+using Microsoft.Xna.Framework;
+
+namespace MonoGame.Utilities.Png
+{
+    public class PngReader
+    {
+        private int width;
+        private int height;
+        private int bitsPerSample;
+        private int bytesPerSample;
+        private int bytesPerPixel;
+        private int bytesPerScanline;
+        private IList<PngChunk> chunks;
+        private IList<PngChunk> dataChunks;
+        private ColorType colorType;
+        private Palette palette;
+        private Texture2D texture;
+        private Color[] data;
+        
+        public PngReader()
+        {
+            chunks = new List<PngChunk>();
+            dataChunks = new List<PngChunk>();
+        }
+
+        public Texture2D Read(Stream inputStream, GraphicsDevice graphicsDevice)
+        {
+            byte[] signature = new byte[8];
+            inputStream.Read(signature, 0, 8);
+
+            if (signature.EqualsByElement(HeaderChunk.PngSignature) == false)
+            {
+                throw new Exception("File does not have PNG signature.");
```

Comment:
Ideally we should have a separate `PngReader.IsPng(Stream)` method so that `Texture2D.FromStream` can eventually test for PNG first.